File: c/sorts/mergesort.c
#include <stdio.h>
#include <stdlib.h> // malloc
#include <string.h> // memcpy
#include <time.h>

#define RANGE 100
#define SIZE 10

int merge(int *a, int a_len, int *b, int b_len, int *out)
{
    int i, j, k;
    i = j = k = 0;

    while (i < a_len && j < b_len) {
        out[k++] = (a[i] < b[j])
            ? a[i++]
            : b[j++];
    }

    while (i < a_len) out[k++] = a[i++];
    while (j < b_len) out[k++] = b[j++];
}

void sort(int *arr, int *tmp, int len)
{
    if (len <= 1) return;

    int mid = len / 2;

    sort(tmp, arr, mid);
    sort(tmp + mid, arr + mid, len - mid);

    merge(tmp, mid, tmp + mid, len - mid, arr);
}

void merge_sort(int *arr, int len)
{
    int *tmp = malloc(sizeof(int) * len);
    memcpy(tmp, arr, sizeof(int) * len);

    sort(arr, tmp, len);

    free(tmp);
}

int sorted(int *buf, int len)
{
    int i;
    for (i = 1; i < len; i++) {
        if (buf[i - 1] > buf[i]) return 0;
    }

    return 1;
}

int main()
{
    srand(time(NULL));

    int i, x[SIZE];

    for (i = 0; i < SIZE; i++) {
        x[i] = rand() % RANGE + 1; // 1 to RANGE
    }

    printf("Before: ");
    for (i = 0; i < SIZE; i++) printf("%d ", x[i]);
    putchar('\n');

    merge_sort(x, SIZE);

    printf("After: ");
    for (i = 0; i < SIZE; i++) printf("%d ", x[i]);
    putchar('\n');

    printf("Sorted: %s\n", sorted(x, SIZE) ? "Yes" : "No");

    return 0;
}

File: java/sorts/MergeSort.java
public class MergeSort {

    public static void merge(Comparable[] arr, Comparable[] tmp, int l, int m, int h)
    {
        // copy current order into temp array.
        for (int i = l; i <= h; i++) {
            tmp[i] = arr[i];
        }

        int i = l, j = m + 1;

        for (int k = l; k <= h; k++) {
            if (i > m) {
                // left complete.
                arr[k] = tmp[j++];
            } else if (j > h) {
                // right complete.
                arr[k] = tmp[i++];
            } else if (tmp[j].compareTo(tmp[i]) < 0) {
                // right < left.
                arr[k] = tmp[j++];
            } else {
                // left < right.
                arr[k] = tmp[i++];
            }
        }
    }

    public static void sort(Comparable[] arr, Comparable[] tmp, int l, int h)
    {
        // return if 0..1
        if (l >= h) {
            return;
        }

        int m = l + (h - l) / 2;

        sort(arr, tmp, l, m);     // left.
        sort(arr, tmp, m + 1, h); // right.

        merge(arr, tmp, l, m, h);
    }

    public static void sort(Comparable[] arr)
    {
        Comparable[] tmp = new Comparable[arr.length];
        sort(arr, tmp, 0, arr.length - 1);
    }

    public static void main(String[] args)
    {
        String[] alphabet = new String[] { "b", "a", "d", "c" };

        System.out.print("Before: ");
        for (String a : alphabet) System.out.print(a + " ");
        System.out.println();

        sort(alphabet);

        System.out.print("After: ");
        for (String a : alphabet) System.out.print(a + " ");
        System.out.println();
    }

}